frappe/email/receive.py

```diff
@@ -149,37 +153,86 @@ def get_messages(self):
 			else:
 				self.pop.quit()
 
-		return self.latest_messages
+		out = { "latest_messages": self.latest_messages }
+		if self.settings.use_imap:
+			out.update({
+				"uid_list": uid_list,
+				"seen_status": self.seen_status,
+				"unique_id_list": self.unique_id_list
+			})
+
+		return out
 
 	def get_new_mails(self):
 		"""Return list of new mails"""
 		if cint(self.settings.use_imap):
-			self.imap.select("Inbox")
-			response, message = self.imap.uid('search', None, "UNSEEN")
+			self.check_imap_uidvalidity()
+
+			self.imap.select("Inbox", readonly=True)
+			response, message = self.imap.uid('search', None, self.settings.email_sync_rule)
 			email_list =  message[0].split()
 		else:
 			email_list = self.pop.list()[1]
 
 		return email_list
 
+	def check_imap_uidvalidity(self):
+		# compare the UIDVALIDITY of email account and imap server
+		uid_validity = self.settings.uid_validity
+
+		responce, message = self.imap.status("Inbox", "(UIDVALIDITY UIDNEXT)")
+		current_uid_validity = self.parse_imap_responce("UIDVALIDITY", message[0])
+		if not current_uid_validity:
+			frappe.throw(_("Can not find UIDVALIDITY in imap status response"))
+
+		uidnext = int(self.parse_imap_responce("UIDNEXT", message[0]) or "1")
+		frappe.db.set_value("Email Account", self.settings.email_account, "uidnext", uidnext)
+
+		if not uid_validity or uid_validity != current_uid_validity:
+			# uidvalidity changed & all email uids are reindexed by server
+			frappe.db.sql("""update `tabCommunication` set uid=-1 where communication_medium='Email'
+				and email_account='{email_account}'""".format(email_account=self.settings.email_account))
+			frappe.db.sql(""" update `tabEmail Account` set uidvalidity='{uidvalidity}', uidnext={uidnext} where
+				name='{email_account}'""".format(uidvalidity=current_uid_validity, uidnext=uidnext, email_account=self.settings.email_account))
+
+			from_uid = 1 if uidnext < 101 or (uidnext - 100) < 1 else uidnext - 100
+			# sync last 100 email
+			self.settings.email_sync_rule = "UID {}:{}".format(from_uid, uidnext)
+		
+		elif uid_validity == current_uid_validity:
+			return
+
+	def parse_imap_responce(self, cmd, responce):
+		pattern = r"(?<={cmd} )[0-9]*".format(cmd=cmd)
+		match = re.search(pattern, responce, re.U | re.I)
+		if match:
+			return match.group(0)
+		else:
+			return None
+
 	def retrieve_message(self, message_meta, msg_num=None):
 		incoming_mail = None
 		try:
 			self.validate_message_limits(message_meta)
 
 			if cint(self.settings.use_imap):
-				status, message = self.imap.uid('fetch', message_meta, '(RFC822)')
-				self.latest_messages.append(message[0][1])
+				status, message = self.imap.uid('fetch', message_meta, '(RFC822 BODY.PEEK[HEADER] FLAGS)')
```

Comment:
@RobertSchouten,

BODY.PEEK[HEADER] is used to generate the fingerprint, RFC822 fetches the email body but not the header also used the Flags in single call to avoid executing multiple imap fetch commands